ideascube/forms.py

```diff
@@ -52,3 +60,37 @@ def save(self, *args, **kwargs):
         user.is_staff = True
         user.save()
         return user
+
+
+class UserImportForm(forms.Form):
+    source = forms.FileField(required=True)
+
+    def save(self):
+        source = self.cleaned_data['source']
+        users = []
+        errors = []
+        rows = csv.DictReader(source.read().decode().splitlines())
+        for idx, row in enumerate(rows):
+            serial = row.get('serial')
+            if not serial:
+                continue
+            data = {}
+            for field in User._meta.fields:
+                try:
+                    data[field.name] = row[field.name]
+                except KeyError:
+                    continue
+                if isinstance(field, models.CharField):
+                    data[field.name] = to_unicode(data[field.name])
+
+            instance = User.objects.filter(serial=serial).first()
+            form = UserForm(data=data, instance=instance)
+            if form.is_valid():
+                user = form.save()
+                users.append(user)
+            else:
+                reason = ', '.join('{}: {}'.format(k, v.as_text())
+                                   for k, v in form.errors.items())
+                errors.append(_('Invalid row at line {id}: {reason}'.format(
+                    id=idx + 1, reason=reason)))
```

Comment:
Yeah, agreed.
I've been a bit disturbed by both the need to get the serial out of the data in order to query for the instance, and for the need to convert to unicode. But I've refactored now and it's much clearer and shorter I think. Thanks :)